extmod/modssl_mbedtls.c

```diff
@@ -138,6 +171,51 @@ STATIC NORETURN void mbedtls_raise_error(int err) {
     #endif
 }
 
+STATIC void ssl_async_handshake(mp_obj_ssl_socket_t *sslsock, int *errcode) {
+
+    if (
+        #if MBEDTLS_VERSION_NUMBER >= 0x03000000
+        (*errcode < 0) && (mbedtls_ssl_is_handshake_over(&sslsock->ssl) == 0) && (*errcode != MBEDTLS_ERR_SSL_CONN_EOF)
+        #else
+        (*errcode < 0) && (*errcode != MBEDTLS_ERR_SSL_CONN_EOF)
+        #endif
+        ) {
+        // Async handshake is done by mbdetls_ssl_read/write
+        // if return code is MBEDTLS_ERR_XX (i.e < 0) and handshake is not done due to
+        // handshake failure notify peer
+        // with proper error code and raise mp error with mbedtls_raise_error
+        uint8_t alert = MBEDTLS_SSL_ALERT_MSG_HANDSHAKE_FAILURE;
+        uint32_t flags = 0;
+        uint ret = 0;
+        if (*errcode == MBEDTLS_ERR_X509_CERT_VERIFY_FAILED) {
+            // The certificate may have been rejected for several reasons.
+            flags = mbedtls_ssl_get_verify_result(&sslsock->ssl);
+
+        } else if (*errcode == MBEDTLS_ERR_SSL_NO_CLIENT_CERTIFICATE) {
+            /* alert = MBEDTLS_SSL_ALERT_MSG_CERT_REQUIRED; tlsv1.3 */
+            mbedtls_ssl_send_alert_message(&sslsock->ssl, MBEDTLS_SSL_ALERT_LEVEL_FATAL,
+                alert);
+        }
+
+        if (*errcode == MBEDTLS_ERR_X509_CERT_VERIFY_FAILED) {
+            char xcbuf[512];
```

Comment:
Does this need to be 512 bytes?  That's quite a lot of stack space.  Can it be 64 without losing information?  How about 128?